param.go

```diff
@@ -326,3 +342,52 @@ func (po paramObject) Build(c *Container) (reflect.Value, error) {
 	}
 	return dest, nil
 }
+
+// paramGroupedSlice is a param which produces a slice of values with the same
+// group name.
+type paramGroupedSlice struct {
+	// Name of the group as specified in the `group:".."` tag.
+	Group string
+
+	// Type of the slice.
+	Type reflect.Type
+}
+
+// newParamGroupedSlice builds a paramGroupedSlice from the provided type with
+// the given name.
+//
+// The type MUST be a slice type.
+func newParamGroupedSlice(group string, t reflect.Type) paramGroupedSlice {
+	return paramGroupedSlice{Group: group, Type: t}
+}
+
+func (pt paramGroupedSlice) Build(c *Container) (reflect.Value, error) {
+	k := key{group: pt.Group, t: pt.Type.Elem()}
+
+	for _, n := range c.producers[k] {
+		if err := n.Call(c); err != nil {
+			// TODO: better message
+			return _noValue, errWrapf(err, "failed to build %v", k)
+		}
+	}
+
+	items := c.groups[k]
+
+	// shuffle the list so users don't rely on the ordering of grouped values
```

Comment:
ehhhhhhhhhhhh idk, does this really do anything? maybe people aren't allowed to rely on the ordering, but there's some optimization they take into account when things are "ordered" in the way they generally expect? Think golang maps, which aren't ordered, but when you iterate the iteration order is generally the insertion order. It just feels like shuffling is eh.

Reply:
This is very deliberate. Groups are unordered so you don't get to rely even on the implicit ordering.

Also, Go map iteration order is not insertion order. It's intentionally randomized by the runtime.

https://play.golang.org/p/x1Covh7eIv
https://nathanleclaire.com/blog/2014/04/27/a-surprising-feature-of-golang-that-colored-me-impressed/

Reply:
Go map ordering is explicitly randomized. https://play.golang.org/p/tKEWlW8Fiv

Reply:
This mitigates a common problem in C. You’ll often only test short-lived programs and use pointers as hashmap keys. The first run will tend to lay the map out sequentially as malloc pioneers the heap. It’ll work well enough on the first run. You’ll discover new behaviors in production (too late) when the heap turns to Swiss cheese.